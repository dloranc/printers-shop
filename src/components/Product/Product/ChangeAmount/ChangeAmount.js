import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { ProductLabel } from './../ProductLabel/ProductLabel';

const AmountLabel = styled(ProductLabel)`
    margin-top: 2em;
`;

export const AmountInput = styled.input`
    margin: 0.3em 0;
    padding: 0.3em 0.5em;
    width: 100%;
`;

export class ChangeAmount extends Component {
  static propTypes = {
    amount: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]).isRequired,
    onChange: PropTypes.func.isRequired
  }

  state = {
    amount: 1
  }

  render() {
    return (
      <div>
        <AmountLabel
          as="label"
          htmlFor={'amount'}
        >
          Amount of items to order:
        </AmountLabel>

        <AmountInput
          type="number"
          name="amount"
          id={'amount'}
          value={this.props.amount}
          onChange={this.props.onChange}
        />
      </div>
    );
  }
}

export default ChangeAmount;
