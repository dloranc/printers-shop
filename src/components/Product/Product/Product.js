import React, { Component } from 'react'
import styled from 'styled-components';

const ProductCard = styled.div`
    width: 100%;
    padding: 1em;
    background: #fefefe;
    color: #333;
    box-shadow: 0 0 0.4em #ccc;
`;

const Title = styled.h2`
    font-weight: lighter;
    margin-bottom: 0.5em;
`;

const ProductRow = styled.div`
    color: #666;
`;

const ProductLabel = styled.span`
    color: black;
    font-weight: bold;
`;

const AmountLabel = styled(ProductLabel)`
    margin-top: 2em;
`;

const AmountInput = styled.input`
    margin: 0.3em 0;
    padding: 0.3em 0.5em;
    width: 100%;
`;

const AddToCartButton = styled.button`
    display: block;
    margin: 0 auto;
    margin-top: 1em;
    padding: 0.6em 1em;
    color: white;
    background: #007bff;
    border: 0;

    &[disabled] {
        background: #eee;
    }

    :hover {
        background: #118cff;
    }
`;

export class Product extends Component {
    state = {
        amount: 1,
    }

    inStock = () => {
        const inStock = this.props.inStock;

        if (inStock > 100) {
            return 'full supply';
        }

        if (inStock > 10) {
            return 'medium supply';
        }

        if (inStock > 0) {
            return 'last pieces';
        }

        return 'not available';
    }

    addToCart = () => {
        if (this.state.amount > this.props.inStock) {
            window.alert('Sorry, but your requested amount of this product exceeds our supply!');
        } else {
            if (this.state.amount > 0) {
                // TODO: add to cart (Redux store)
                window.alert('Added to cart!');
            }
        }
    }

    handleChange = event => {
        const value = event.target.value;

        if (value < 0) {
            this.setState({ amount: 0 });
        } else {
            this.setState({ amount: value });
        }
    }

    render() {
        return (
            <ProductCard>
                <Title>{this.props.type} {this.props.name}</Title>

                <ProductRow>
                    <ProductLabel>Price per unit:</ProductLabel> ${this.props.price}
                </ProductRow>

                <ProductRow>
                    <ProductLabel>In stock:</ProductLabel> {this.inStock()}
                </ProductRow>

                <AmountLabel
                    as="label"
                    htmlFor={'amount-' + this.props.id}
                >
                    Amount of items to order:
                </AmountLabel>

                <AmountInput
                    type="number"
                    name="amount"
                    id={'amount-' + this.props.id}
                    value={this.state.amount}
                    onChange={this.handleChange}
                />

                <ProductRow>
                    <ProductLabel>Price:</ProductLabel> ${this.state.amount * this.props.price}
                </ProductRow>

                <AddToCartButton disabled={this.state.amount === 0} onClick={this.addToCart}>Add to cart</AddToCartButton>
            </ProductCard>
        )
    }
}

export default Product;
