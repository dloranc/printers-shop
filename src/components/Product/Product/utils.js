export const inStock = inStock => {
  if (inStock > 100) {
    return 'full supply';
  }

  if (inStock > 10) {
    return 'medium supply';
  }

  if (inStock > 0) {
    return 'last pieces';
  }

  return 'not available';
}
