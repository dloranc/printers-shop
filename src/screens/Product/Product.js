import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter, Redirect } from 'react-router-dom';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { Fetcher } from '../../components/Fetcher/Fetcher';
import ProductView from '../../components/Product/Product/View/View';

class ScreensProduct extends Component {
  static propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    match: PropTypes.object.isRequired
  }

  render() {
    if (this.props.isAuthenticated) {
      const productId = this.props.match.params.id;

      const url = `http://localhost:4000/products/${productId}`;

      return (
        <>
          <Helmet>
            <title>Product - Printers Shop</title>
          </Helmet>

          <Fetcher url={url}>
            {({ data, isLoading, error }) => {
              if (isLoading) {
                return <p>Loading...</p>;
              }

              if (error) {
                if (error.response.status === 404) {
                  return <Redirect to='/404'/>;
                }

                return <p>{error.message}</p>;
              }

              if (!isLoading) {
                return <ProductView product={data}/>;
              }
            }}
          </Fetcher>
        </>
      );
    }

    return <Redirect to="/"></Redirect>;
  }
}

const mapStateToProps = (state) => {
  return {
    isAuthenticated: state.user.isAuthenticated
  };
};

export default compose(
  withRouter,
  connect(mapStateToProps)
)(ScreensProduct);
