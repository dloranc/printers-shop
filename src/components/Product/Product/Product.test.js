import { Product } from './Product';
import ChangeAmount from './ChangeAmount/ChangeAmount';

const defaultProps = {
  id: 'product-1',
  name: 'ZX3',
  type: 'Printer',
  price: 400,
  inStock: 10,
  addToCart: () => {}
};

// eslint-disable-next-line no-undef
const setup = buildSetup(Product, defaultProps);

describe('A product compoment', () => {
  it('contain a "Printer ZX3" title', () => {
    const { wrapper } = setup();

    expect(wrapper.text()).toContain('Printer ZX3');
  });

  it('contain price per unit', () => {
    const { wrapper } = setup();

    expect(wrapper.text()).toContain('Price per unit: $400');
  });

  it('contain "not available" text if there are 0 products in stock', () => {
    const { wrapper } = setup({ inStock: 0 });

    expect(wrapper.text()).toContain('not available');
  });

  it('contain "last pieces" text if there are 10 products in stock', () => {
    const { wrapper } = setup();

    expect(wrapper.text()).toContain('last pieces');
  });

  it('contain "medium supply" text if there are 11 products in stock', () => {
    const { wrapper } = setup({ inStock: 11 });

    expect(wrapper.text()).toContain('medium supply');
  });

  it('contain "full supply" text if there are 101 products in stock', () => {
    const { wrapper } = setup({ inStock: 101 });

    expect(wrapper.text()).toContain('full supply');
  });

  it('increase a price when the amount of items to order increases', () => {
    const { wrapper } = setup();

    wrapper.find(ChangeAmount).simulate('change', { target: { value: 2 } });

    expect(wrapper.text()).toContain('Price: $800');
  });
});
