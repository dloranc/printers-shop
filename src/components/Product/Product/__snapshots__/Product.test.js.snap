// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A product compoment renders ZX3 Printer with price $400 and last pieces in stock 1`] = `
<div
  className="sc-bdVaJa gDmiUu"
>
  <h2
    className="sc-bwzfXH gyKPGc"
  >
    Printer
     
    ZX3
  </h2>
  <div
    className="sc-htpNat laTiSR"
  >
    <span
      className="sc-bxivhb hlixEs"
    >
      Price per unit:
    </span>
     $
    400
  </div>
  <div
    className="sc-htpNat laTiSR"
  >
    <span
      className="sc-bxivhb hlixEs"
    >
      In stock:
    </span>
     
    last pieces
  </div>
  <label
    className="sc-bxivhb sc-ifAKCX jCMYwx"
    htmlFor="amount-product-1"
  >
    Amount of items to order:
  </label>
  <input
    className="sc-EHOje jTIRCN"
    id="amount-product-1"
    name="amount"
    onChange={[Function]}
    type="number"
    value={1}
  />
  <div
    className="sc-htpNat laTiSR"
  >
    <span
      className="sc-bxivhb hlixEs"
    >
      Price:
    </span>
     $
    400
  </div>
  <button
    className="sc-bZQynM hxtQYt"
    disabled={false}
    onClick={[Function]}
  >
    Add to cart
  </button>
</div>
`;
